packages/astro/src/actions/runtime/middleware.ts

```diff
@@ -6,14 +6,18 @@ import {
 } from '../../core/errors/errors-data.js';
 import { AstroError } from '../../core/errors/errors.js';
 import { defineMiddleware } from '../../core/middleware/index.js';
-import { formContentTypes, getAction, hasContentType } from './utils.js';
-import { getActionQueryString } from './virtual/shared.js';
+import { formContentTypes, hasContentType } from './utils.js';
+import {
+	type SafeResult,
+	type SerializedActionResult,
+	serializeActionResult,
+} from './virtual/shared.js';
+import { getAction } from './virtual/get-action.js';
 
 export type Locals = {
 	_actionsInternal: {
-		getActionResult: APIContext['getActionResult'];
-		callAction: APIContext['callAction'];
-		actionResult?: ReturnType<APIContext['getActionResult']>;
+		actionResult: SerializedActionResult;
+		actionName: string;
```

Comment:
Moved the `getActionResult` constructor to `render-context`. This ensures only serializable values are sent through `locals` for edge middleware support. It also simplified our code in the process by removing that `nextLocalsStub` utility!